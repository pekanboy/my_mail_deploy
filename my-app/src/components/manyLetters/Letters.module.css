.zero {
    transform: translateX(-50%);
    position: fixed;
    top: 128px;
    left: 50%;

    font-family: "Roboto", sans-serif;
    font-style: normal;
    font-weight: normal;
    font-size: 22px;
    line-height: 43px;
    align-items: center;
    text-align: center;
}

.letters {
    margin-top: 4px;
}