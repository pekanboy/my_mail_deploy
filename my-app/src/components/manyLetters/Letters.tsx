import React from 'react';
import {ILetter} from '../../interfaces/ILetter';
import {Letter} from '../Letter/Letter';
import styles from './Letters.module.css';

export interface LettersProps {
    letters: ILetter[]
    selectHandler: (id :number) => void
}

export const Letters: React.FC<LettersProps> = ({letters, selectHandler}) => {
    if (!letters.length) {
        return (
            <div className={styles.zero}>
                У вас нет входящих писем
            </div>
        );
    }

    return (
        <div className={styles.letters}>
            {
                letters?.map((letter) => {
                    return (
                        <Letter
                            key={letter.id.toString()}
                            letter={letter}
                            checkHandler={selectHandler}
                        />
                    );
                },
                )}
        </div>
    );
};
