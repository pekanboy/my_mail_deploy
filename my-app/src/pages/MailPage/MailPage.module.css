.container {
    max-width: 1040px;
    margin: 0 auto;
    position: relative;
    display: flex;
    flex-direction: column;
}

.nav {
    display: flex;
}

.title {
    position: relative;

    width: 150px;
    height: 80px;

    left: 50%;
    transform: translateX(-50%);
    color: #FF0000;

    font-family: "Roboto Light", sans-serif;
    font-style: normal;
    font-weight: normal;
    font-size: 36px;
    line-height: 43px;
    align-items: center;
    text-align: center;
}

.crimson {
    color: #3A0000;
}

body {
    background-image: url("../../static/bg.jpeg");
}
