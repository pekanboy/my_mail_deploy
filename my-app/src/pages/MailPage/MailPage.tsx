import React, {useCallback, useEffect, useState} from 'react';
import styles from './MailPage.module.css';
import {SelectCheckbox} from '../../components/SelectCheckbox/SelectCheckbox';
import {DeleteButton} from '../../components/Buttons/DeleteButton/DeleteButton';
import {ILetter} from '../../interfaces/ILetter';
import {Letters} from '../../components/manyLetters/Letters';

export interface MainPageProps {
    emails: ILetter[]
}

export const MailPage: React.FC<MainPageProps> = ({emails}) => {
    const [isAllSelected, setIsAllSelected] = useState<boolean>(false);
    const [letters, setLetters] = useState<ILetter[]>(emails);

    useEffect(() => {
        setLetters((prevLetters) => {
            return prevLetters.map((letter) => {
                letter.isSelected = isAllSelected;
                return letter;
            });
        });
    }, [isAllSelected]);

    const allCheckHandler = useCallback(() => {
        setIsAllSelected((prev) => !prev);
    }, []);

    const oneCheckHandler = useCallback((id: number) => {
        setLetters(letters.map((letter) => {
            if (letter.id !== id) {
                return letter;
            }

            letter.isSelected = !letter.isSelected;
            return letter;
        }));
    }, [letters]);

    const deleteHandler = useCallback(() => {
        setLetters((prevLetters) => {
            return prevLetters.filter((letter) => !letter.isSelected);
        });
    }, []);


    return (
        <div className={styles.container}>
            <div className={styles.title}>
                <span className={styles.crimson}>My</span>Mail
            </div>
            <div className={styles.nav}>
                <SelectCheckbox
                    checked={isAllSelected}
                    checkHandler={allCheckHandler}/>
                <DeleteButton
                    clickHandler={deleteHandler}/>
            </div>
            <Letters
                letters={letters}
                selectHandler={oneCheckHandler}/>
        </div>
    );
};
