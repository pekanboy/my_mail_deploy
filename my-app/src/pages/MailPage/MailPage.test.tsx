import React from 'react';
import {configure, shallow, ShallowWrapper} from 'enzyme';
import {MailPage, MainPageProps} from './MailPage';
import Adapter from 'enzyme-adapter-react-16';
import {LettersProps} from '../../components/manyLetters/Letters';
import {DeleteButtonProps}
    from '../../components/Buttons/DeleteButton/DeleteButton';

configure({adapter: new Adapter(), disableLifecycleMethods: false});

const testData: MainPageProps = {
    emails: [
        {
            'id': 1,
            'avatar': 'http://i.pravatar.cc/300?u=1',
            'sender': 'Бединге',
            'theme': 'Carolina Hurricanes',
            'snippet': 'Люди всё время меня спрашивают: ' +
                'знаю ли я Тайлера Дёрдена?',
        },
        {
            'id': 2,
            'avatar': 'http://i.pravatar.cc/300?u=2',
            'sender': 'Лиаторп',
            'theme': 'Columbus Blue Jackets',
            'snippet': 'Шесть месяцев я не мог спать. ' +
                'Когда у тебя бессонница — всё нереально; ' +
                'всё очень далеко от тебя, всё это — копия, ' +
                'снятая с копии, которая в свою очередь снята с копии.',
        },
        {
            'id': 3,
            'avatar': 'http://i.pravatar.cc/300?u=3',
            'sender': 'Хемнэс',
            'theme': 'New Jersey Devils',
            'snippet': 'Когда-то мы зачитывались ' +
                'порнографией — теперь каталогами IKEA.',
        }],
};

describe('test mail page', () => {
    it('test data snapshot match', () => {
        const wrapper: ShallowWrapper = shallow(<MailPage {...testData} />);

        expect(wrapper).toMatchSnapshot();
    });

    it('delete one letters test', () => {
        const wrapper: ShallowWrapper = shallow(<MailPage {...testData} />);

        const LettersProps = wrapper.find('Letters').props() as LettersProps;
        LettersProps.selectHandler(1);

        expect(
            (wrapper.find('Letters').props() as LettersProps)
                .letters
                .filter((letter) => letter.isSelected)
                .length,
        ).toBe(1);

        const DeleteButtonProps = wrapper
            .find('DeleteButton').props() as DeleteButtonProps;
        DeleteButtonProps.clickHandler();

        expect(
            (wrapper.find('Letters').props() as LettersProps)
                .letters
                .length,
        ).toBe(testData.emails.length - 1);
    });

    // TODO: Проблема в том, что тест не дожидается ререндера компонента
    //  и expect срабатывает быстрее, чем useEffect
    // it('select all test', () => {
    //     const wrapper: ShallowWrapper = shallow(<MailPage {...testData} />,)
    //
    //     const SelectCheckboxProps = wrapper
    //     .find('SelectCheckbox')
    //     .props() as SelectCheckboxProps;
    //     SelectCheckboxProps.checkHandler();
    //
    //     expect(
    //         (wrapper.find('Letters').props() as LettersProps)
    //             .letters
    //             .filter(letter => letter.isSelected)
    //             .length
    //     ).toBe(testData.emails.length);
    // });
});
