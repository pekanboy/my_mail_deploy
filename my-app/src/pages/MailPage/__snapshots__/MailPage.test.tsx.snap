// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test mail page test data snapshot match 1`] = `
<div
  className="container"
>
  <div
    className="title"
  >
    <span
      className="crimson"
    >
      My
    </span>
    Mail
  </div>
  <div
    className="nav"
  >
    <SelectCheckbox
      checkHandler={[Function]}
      checked={false}
    />
    <DeleteButton
      clickHandler={[Function]}
    />
  </div>
  <Letters
    letters={
      Array [
        Object {
          "avatar": "http://i.pravatar.cc/300?u=1",
          "id": 1,
          "sender": "Бединге",
          "snippet": "Люди всё время меня спрашивают: знаю ли я Тайлера Дёрдена?",
          "theme": "Carolina Hurricanes",
        },
        Object {
          "avatar": "http://i.pravatar.cc/300?u=2",
          "id": 2,
          "sender": "Лиаторп",
          "snippet": "Шесть месяцев я не мог спать. Когда у тебя бессонница — всё нереально; всё очень далеко от тебя, всё это — копия, снятая с копии, которая в свою очередь снята с копии.",
          "theme": "Columbus Blue Jackets",
        },
        Object {
          "avatar": "http://i.pravatar.cc/300?u=3",
          "id": 3,
          "sender": "Хемнэс",
          "snippet": "Когда-то мы зачитывались порнографией — теперь каталогами IKEA.",
          "theme": "New Jersey Devils",
        },
      ]
    }
    selectHandler={[Function]}
  />
</div>
`;
